predicators/nsrt_learning/strips_learning/pnad_search_learner.py

```diff
@@ -0,0 +1,361 @@
+    GeneralToSpecificSTRIPSLearner
+from predicators.structs import GroundAtom, LowLevelTrajectory, \
+    ParameterizedOption, PartialNSRTAndDatastore, Predicate, Segment, Task, \
+    _GroundSTRIPSOperator
+
+
+class _PNADSearchOperator(abc.ABC):
+    """An operator that proposes successor sets of PNAD sets."""
+
+    def __init__(self, trajectories: List[LowLevelTrajectory],
+                 train_tasks: List[Task], predicates: Set[Predicate],
+                 segmented_trajs: List[List[Segment]],
+                 learner: PNADSearchSTRIPSLearner) -> None:
+        self._trajectories = trajectories
+        self._train_tasks = train_tasks
+        self._predicates = predicates
+        self._segmented_trajs = segmented_trajs
+        self._learner = learner
+
+    @abc.abstractmethod
+    def get_successors(
+        self, pnads: FrozenSet[PartialNSRTAndDatastore]
+    ) -> Iterator[FrozenSet[PartialNSRTAndDatastore]]:
+        """Generate zero or more successor PNAD sets."""
+        raise NotImplementedError("Override me!")
+
+
+class _BackChainingPNADSearchOperator(_PNADSearchOperator):
+    """An operator that uses backchaining to propose a new PNAD set."""
+
+    def get_successors(
+        self, pnads: FrozenSet[PartialNSRTAndDatastore]
+    ) -> Iterator[FrozenSet[PartialNSRTAndDatastore]]:
+        pnads_list = list(pnads)
+        pnads_list = sorted(pnads_list)
+        uncovered_segment = self._get_first_uncovered_segment(pnads_list)
+        if uncovered_segment is not None:
+            # We will need to induce an operator to cover this
+            # segment, and thus it must have some necessary add effects.
+            new_pnad = self._learner.spawn_new_pnad(uncovered_segment)
+            ret_pnads_list = self._append_new_pnad_and_keep_effects(
+                new_pnad, pnads_list)
+            ret_pnads = frozenset(ret_pnads_list)
+            yield ret_pnads
+
+    def _append_new_pnad_and_keep_effects(
+        self, new_pnad: PartialNSRTAndDatastore,
+        current_pnads: List[PartialNSRTAndDatastore]
+    ) -> List[PartialNSRTAndDatastore]:
+        """Given some newly-created PNAD and a set of existing PNADs, correctly
+        repartition data amongst all these PNADs and induce keep effects for
+        the newly-created PNAD.
+
+        Return the final set of all PNADs (current, newly-created, and
+        ones with keep effects).
+        """
+        new_pnads = current_pnads + [new_pnad]
+        # We first repartition data and ensure delete and ignore
+        # effects for the newly-created PNAD are correct.
+        new_pnads = self._learner.recompute_pnads_from_effects(new_pnads)
+        # Ensure that the unnecessary keep effs sub and poss
+        # keep effects are both cleared before backchaining. This is
+        # important because we will be inducing keep effects after this
+        # backchaining.
+        for pnad in new_pnads:
+            pnad.poss_keep_effects.clear()
```

Comment:
it looks like you do this clear before all calls of the method below. how about we just make the method below called `clear_keep_effs` and it calls `pnad.poss_keep_effects.clear()` and then also does the current logic for the subs